Method rename support (initial implementation)

soam-stripe opened the pull request:
Implements a first cut of method rename support, building on the constant rename work. This implementation supports subclasses but doesn't handle union types, mixins, and probably more cases.

The goal of this change is to enable method renames in experimental mode and gather some user feedback about which missing bits are most important.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

We have lots of user demand for refactoring features, and we think method renames would one of the useful ones.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

jvilk-stripe reviewed (changes requested):
This is looking good! I had one big correctness note and one big maintainability note. The rest are small nits.

I haven't reviewed the test coverage; I'll do that once the other comments are addressed.

soam-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -29,37 +30,135 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
-            return JSONNullObject();
+    vector<core::SymbolRef> symbolsToRename;
+    if (symbolData->isMethod()) {
+        // We have to check for methods as part of a class hierarchy: Follow superClass() links till we find the root;
```
Excellent point! Will fix.

soam-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -29,37 +30,135 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
+    auto super = s.data(gs)->superClass();
+    if (super.exists()) {
+        if (updateSubclass(gs, root, super, isSubclass, visited, subclasses)) {
+            subclasses.push_back(s);
```
I don't see how it duplicates results? The function can't ever do line 59 and this line for the same class because it tracks `visited`.  If we move out the vector from this function we'll add a factor of `class hierarchy length` to the overall time complexity, I think? Because we'll have to walk up the path from class to root for every class we visit, e.g. if we have `A < B < C`, we'll have to walk the B->C link twice if we happen to visit A before B in the outer loop.

soam-stripe commented:
Okay I've addressed all the comments except the ones about updateSubclass. 

ptal @jvilk-stripe

soam-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -29,37 +30,135 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
+// simple ReplaceAll.
+std::string replaceLastDotted(std::string_view input, std::string_view originalName, std::string_view newName) {
+    if (absl::StrContains(input, ".")) {
+        std::vector<std::string> dotted = absl::StrSplit(input, ".");
```
(I think I can't use auto here because there are a few return-type overloads of absl::StrSplit)

soam-stripe commented:
Thanks for the review @jvilk-stripe, the `isSubclass` function is updated too.

jvilk-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -114,15 +241,24 @@ unique_ptr<ResponseMessage> RenameTask::runRequest(LSPTypecheckerDelegate &typec
             // Only supports rename requests from constants and class definitions.
             if (auto constResp = resp->isConstant()) {
+                // Sanity check the text.
+                if (islower(params->newName[0])) {
```
This check is also going to be needed if `resp->isDefinition()` and the symbol `isClassOrModule()`.

This reminds me that _method_ renames have to be lower case. You can choose if you want to add that check now or later. :)

jvilk-stripe commented on test/testdata/lsp/rename/method_attr.rb:
```diff
@@ -0,0 +1,10 @@
+
+f = Foo.new
+f.bar
+# ^ apply-rename: [A] newName: baz
```
Note that this won't completely work since `attr_reader :bar` also creates a `@bar` in the class, which the class assigns to to populate the reader. But that's an acceptable limitation for now; we'd need the ability 1) to recognize it's an `attr_reader` and 2) to rename the field.

soam-stripe replied:
Yeah... good point. Do you think we should explicitly check for this and reject the rename? Or is it enough just to document this as a current limitation?

jvilk-stripe replied:
I'm OK with it as a current limitation.

jvilk-stripe replied:
Updating the field references is much easier than the method references, as the field can only be referenced within the class methods.

jvilk-stripe commented on test/testdata/lsp/rename/method_class_hierarchy.rb:
```diff
@@ -0,0 +1,30 @@
+
+class C < Base
+  def foo
+    # this one should not be renamed!
```
Nice

jvilk-stripe reviewed (approved):
The code looks good; I had a few small suggestions.

You might want to add a test for renaming class methods, e.g.:

```ruby
class Foo
  def self.foo
    #       ^ renaming this
  end

  def foo
    # this is a separate method
  end
end

Foo.foo # this is a callsite of self.foo 

Foo.new.foo.# this is a callsite of foo
```

I think it should 'just work'.

I'd also recommend testing that we _reject_ renames on call sites of `new` (e.g. `Foo.new`) and on `initialize` (which `Foo.new` calls).

soam-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -29,48 +30,187 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
+string replaceLastDotted(string_view input, string_view originalName, string_view newName) {
+    vector<string> dotted = absl::StrSplit(input, ".");
+    // TODO: check for exactly one match
+    dotted[dotted.size() - 1] = absl::StrReplaceAll(dotted[dotted.size() - 1], {{originalName, newName}});
```
Your question made me realize this code is just wrong -- in case of a send the loc contains the full method args, so this "split by dots" thing is too simple. It needs to properly parse the function name out of an arbitrary method call. I'll update with a fix.

soam-stripe commented:
Thanks again! Addressed the comments. I'm still not a 100% sure the method parsing stuff is really correct, let me know if you see anything I missed?

PTAL @jvilk-stripe

jvilk-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -29,47 +29,208 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
+        auto source = loc.source(gs);
+        string newsrc;
+        if (absl::StartsWith(source, "def ") || absl::StartsWith(source, "attr_reader") ||
+            absl::StartsWith(source, "attr_accessor")) {
```
`attr_writer` is another.

We can remove this hardcoded logic, btw (maybe in the next PR). `getReferencesToSymbol` actually knows the provenance of these locations from the `QueryResponse` objects it processes and if they are from sends or definitions:

https://github.com/sorbet/sorbet/blob/e932d6f8/main/lsp/LSPTask.cc#L226

Maybe we need to introduce a new method (maybe that `getReferencesToSymbol` calls?) that returns a vector of relevant `QueryResponse` objects.

jvilk-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -29,47 +29,208 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
+        for (string::size_type i = 0; i < send.size(); i++) {
+            if (send[i] == ' ' || send[i] == '(') {
+                methodExpr = send.substr(0, i);
+                args = send.substr(i, send.size() - i);
```
Should this break out of the loop?

Also, does this work on callsites with a receiver? e.g.:

```ruby
  foo.bar(a, b, c)
#  ^^^ this is not the method
#.    ^^^ this is the method
```

is `foo` part of the send loc? It _looks like_ the tests pass, so maybe this isn't a problem.

jvilk-stripe replied:
Nit: Actually, maybe we could replace this with something simpler, like `send.find(methodName)` that verifies that the next token in `send` (if present) isn't alphanumeric? Up to you if you want to do it that way.

soam-stripe replied:
Yes it should break from the loop, will fix that. I wanted to avoid searching for methodName because then we have to handle cases where it appears in the string but isn't actually the method we're looking for (like in the method_names test case)

jvilk-stripe replied:
oh yeah! d'oh. makes sense :)

soam-stripe replied:
And yes, `foo` is part of the send loc for `foo.bar(a,b,c)`. But we only break from the loop on `(` or space, so we get `foo.bar` in methodExpr

jvilk-stripe replied:
After discussing in DM about other corner cases, we will fix this issue in a subsequent PR by surfacing receiver loc information in `SendQueryResponse`s. That will make it easy to exclude the part of the send loc that corresponds to the receiving object.

jvilk-stripe commented on main/lsp/requests/rename.cc:
```diff
@@ -114,15 +268,37 @@ unique_ptr<ResponseMessage> RenameTask::runRequest(LSPTypecheckerDelegate &typec
+                    return response;
+                }
+
+                if (defResp->symbol.data(gs)->isMethod() && isupper(params->newName[0])) {
```
This is missing in `isSend()`.

Maybe we should move new name sanity checks to the renamer or something.

jvilk-stripe replied:
You can test this error in the test suite, btw.

jvilk-stripe reviewed (approved):
Looking good! I had a few minor nits.
